pkg/repository/repository.go

```diff
@@ -44,6 +47,36 @@ func (b BadVCSURLError) Error() string {
 	return fmt.Sprintf("invalid VCS URL supplied %s\n", b.URL)
 }
 
+func getHeadReference(vcsURL string, credentials Credentials) plumbing.ReferenceName {
+	e, err := transport.NewEndpoint(vcsURL)
+	checkErr(err, vcsURL)
+
+	e.User = credentials.Username
```

Comment:
When obtaining a HEAD reference from a public repository (E.g. `https://github.com/google/yamlfmt`) with an expired access token, this function would error out with "authentication required" error. To make this more flexible, it is better to obtain the HEAD reference without authentication first, then if there is an error - obtain the HEAD reference with user-supplied credentials.

Also, it is better to return the error early instead of logging it. E.g. it wouldn't make sense to process any of the statements in this function if the endpoint creation (`e, err := transport.NewEndpoint(vcsURL)`) fails.

`getHeadReference` should look like this:
```go
func getHeadReference(vcsURL string, credentials Credentials) (plumbing.ReferenceName, error) {
	endpoint, err := transport.NewEndpoint(vcsURL)
	if err != nil {
		return "", fmt.Errorf("can't create VCS endpoint: %w", err)
	}

	obtainHEADRef := func(endpoint *transport.Endpoint) (plumbing.ReferenceName, error) {
		gitClient, err := client.NewClient(endpoint)
		if err != nil {
			return "", err
		}

		session, err := gitClient.NewUploadPackSession(endpoint, nil)
		if err != nil {
			return "", err
		}

		info, err := session.AdvertisedReferences()
		if err != nil {
			return "", err
		}

		refs, err := info.AllReferences()
		if err != nil {
			return "", err
		}

		return refs["HEAD"].Target(), nil
	}

	unauthenticatedHEAD, err := obtainHEADRef(endpoint)
	if err != nil {
		const warnMessage = "unable to obtain repo HEAD in an unauthenticated state, retrying with credentials"
		log.WithField("error", err).Warn(warnMessage)

		endpoint.User = credentials.Username
		endpoint.Password = credentials.AccessToken

		return obtainHEADRef(endpoint)
	}

	return unauthenticatedHEAD, nil
}
```

Reply:
> When obtaining a HEAD reference from a public repository (E.g. https://github.com/google/yamlfmt) with an expired access token, this function would error out with "authentication required" error. To make this more flexible, it is better to obtain the HEAD reference without authentication first, then if there is an error - obtain the HEAD reference with user-supplied credentials.

Good find! 

> Also, it is better to return the error early instead of logging it. E.g. it wouldn't make sense to process any of the statements in this function if the endpoint creation (e, err := transport.NewEndpoint(vcsURL)) fails.

Totally, my bad for not propagating the error with the response!